metric/bigtable/src/main/java/com/spotify/heroic/metric/bigtable/api/RowRangeReader.java

```diff
@@ -0,0 +1,237 @@
+                cells.addAll(cellChunks);
+                cellChunks = cells;
+            }
+            List<CellChunk> cellsCarryOver = consumeCells(cellChunks);
+            return new CellChunkHandler(commitCells, consumedCells + cellChunks.size(),
+                incompleteCellData, cellsCarryOver);
+        }
+
+        private List<CellChunk> consumeCells(List<CellChunk> cells) {
+            int nextListStart = 0;
+            for (int i = 0; i < cells.size(); i++) {
+                CellChunk cellChunk = cells.get(i);
+                RowStatusCase rowStatusCase = cellChunk.getRowStatusCase();
+                if (rowStatusCase == RowStatusCase.RESET_ROW && cellChunk.getResetRow()) {
+                    nextListStart = i + 1;
+                } else if (rowStatusCase == RowStatusCase.COMMIT_ROW && cellChunk.getCommitRow()) {
+                    int thisListStart = nextListStart;
+                    nextListStart = i + 1;
+                    if (thisListStart == 0 && nextListStart == cells.size()) {
+                        commitCells.accept(cells);
+                    } else {
+                        commitCells.accept(cells.subList(thisListStart, nextListStart));
+                    }
+                }
+            }
+            return cells.subList(nextListStart, cells.size());
+        }
+    }
+
+    private class StreamObserverImpl implements StreamObserver<ReadRowsResponse> {
+
+        private final String tableUri;
+        private final ReadRowRangeRequest rowRequest;
+        private final ResolvableFuture<?> future;
+        private final CellConsumer cellConsumer;
+        private final AtomicReference<CellChunkHandler> cellChunkHandler;
+        private final AtomicReference<ByteString> lastQualifierRef = new AtomicReference<>();
+
+        private StreamObserverImpl(
+            String tableUri, ReadRowRangeRequest rowRequest, ResolvableFuture<?> future,
+            CellConsumer cellConsumer
+        ) {
+            this.tableUri = tableUri;
+            this.rowRequest = rowRequest;
+            this.future = future;
+            this.cellConsumer = cellConsumer;
+            this.cellChunkHandler = new AtomicReference<>(new CellChunkHandler(this::commit));
+        }
+
+        private void commit(List<CellChunk> cellChunks) {
+            cellConsumer.consume(cellChunks, c -> c.getQualifier().getValue(), CellChunk::getValue);
+        }
+
+        @Override
+        public void onNext(ReadRowsResponse response) {
+            List<CellChunk> cellChunks = response.getChunksList();
+            if (cellChunks.isEmpty()) {
+                log.warn("Got empty cell chunk list");
+                return;
+            }
+            log.trace("Got {} cell chunks", response.getChunksList().size());
+            cellChunkHandler.set(cellChunkHandler.get().consumeChunks(cellChunks));
+            lastQualifierRef.set(cellChunks.get(cellChunks.size() - 1).getQualifier().getValue());
```

Comment:
onNext should definitely not be called in parallel, but potentially from different threads, but memory visibility should be fine without `AtmoicReference`. Will remove.